c_emulator/riscv_callbacks.cpp

```diff
@@ -0,0 +1,118 @@
+#include "riscv_callbacks.h"
+#include "riscv_config.h"
+#include <stdlib.h>
+#include <vector>
+
+std::vector<uint8_t> read_lbits(lbits val)
+{
+  std::vector<uint8_t> data(val.len);
+  mpz_export(data.data(), nullptr, -1, 1, 0, 0, *val.bits);
+  return data;
+}
+
+// Implementations of default callbacks for trace printing.
+//
+// The model assumes that these functions do not change the state of the model.
+unit mem_write_callback(uint64_t addr, uint64_t width, lbits value)
+{
+  if (config_print_mem_access) {
+    std::vector<uint8_t> lbits_data = read_lbits(value);
+    printf("mem[0x%.16lX] <- 0x", addr);
+    for (int i = width - 1; i >= 0; --i)
+      printf("%02hhX", lbits_data[i]);
+    printf("\n");
+  }
+#ifdef RVFI_DII
+  zrvfi_write(addr, width, value);
+#endif
+  return UNIT;
+}
+
+unit mem_read_callback(const char *type, uint64_t addr, uint64_t width,
+                       lbits value)
+{
+  if (config_print_mem_access) {
+    std::vector<uint8_t> lbits_data = read_lbits(value);
+    printf("mem[%s,0x%.16lX] -> 0x", type, addr);
+    for (int i = width - 1; i >= 0; --i)
+      printf("%02hhX", lbits_data[i]);
+    printf("\n");
+  }
+#ifdef RVFI_DII
+  sail_int len;
+  CREATE(sail_int)(&len);
+  CONVERT_OF(sail_int, mach_int)(&len, width);
+  zrvfi_read(addr, len, value);
+  KILL(sail_int)(&len);
+#endif
+  return UNIT;
+}
+
+unit mem_exception_callback(uint64_t addr, uint64_t num_of_exception)
+{
+#ifdef RVFI_DII
+  zrvfi_mem_exception(addr);
+#endif
+  return UNIT;
+}
+
+unit xreg_write_callback(unsigned reg, uint64_t value)
+{
+  if (config_print_reg) {
+    printf("x%d <- 0x%.16lX\n", reg, value);
+  }
+#ifdef RVFI_DII
+  zrvfi_wX(reg, value);
+#endif
+  return UNIT;
+}
+
+unit freg_write_callback(unsigned reg, uint64_t value)
+{
+  // TODO: will only print bits; should we print in floating point format?
+  if (config_print_reg) {
+    printf("f%d <- 0x%.16lX\n", reg, value);
```

Comment:
Using l as the modifier won't work for all systems, should probbaly use the PRIX64 macro from inttypes.h